src/main/java/uk/gov/ons/census/casesvc/service/CaseReceiptService.java

```diff
@@ -17,51 +24,121 @@
     }
 
-    caze.setReceiptReceived(true);
-    caseService.saveCaseAndEmitCaseUpdatedEvent(
-        caze, buildMetadata(causeEventType, ActionInstructionType.CLOSE));
+    Rule rule = rules.get(ruleKey);
+    Case lockedCase = rule.run(caze);
+
+    if (rule.saveAndEmitCase) {
+      caseService.saveCaseAndEmitCaseUpdatedEvent(
+          lockedCase, buildMetadata(causeEventType, ActionInstructionType.CLOSE));
+    }
   }
 
-  private void incrementActualResponseAndSetReceiptedIfAppropriate(
-      Case caze, EventTypeDTO causeEventType) {
-    /*
-      This stops the actualResponses being updated by another thread for another receipt or linking event
-    */
-    Optional<Case> oCase = caseRepository.getCaseAndLockByCaseId(caze.getCaseId());
+  private Key makeRulesKey(Case caze, UacQidLink uacQidLink) {
+    String formType = "HH";
+
+    if (isIndividualQuestionnaireType(uacQidLink.getQid())) {
+      formType = "Ind";
+    } else if (caze.getTreatmentCode().startsWith("CE")) {
+      formType = "CE1";
+    } else if (iscontinuationQuestionnaireTypes(uacQidLink.getQid())) {
+      formType = "Cont";
+    }
+
+    return new Key(caze.getCaseType(), caze.getAddressLevel(), formType);
+  }
+
+  Function<Case, Case> incremenNoReceipt =
+      (caze) -> {
+        Case lockedCase = getCaseAndLockIt(caze.getCaseId());
+        lockedCase.setCeActualResponses(lockedCase.getCeActualResponses() + 1);
+
+        return lockedCase;
+      };
+
+  Function<Case, Case> incrementAndReceipt =
+      (caze) -> {
+        Case lockedCase = incremenNoReceipt.apply(caze);
+
+        if (lockedCase.getCeActualResponses() >= lockedCase.getCeExpectedCapacity()) {
+          lockedCase.setReceiptReceived(true);
+        }
+
+        return lockedCase;
+      };
+
+  Function<Case, Case> receiptCase =
+      (caze) -> {
+        caze.setReceiptReceived(true);
+        return caze;
+      };
+
+  Function<Case, Case> noActionRequired = (caze) -> caze;
+
+  private Case getCaseAndLockIt(UUID caseId) {
+    Optional<Case> oCase = caseRepository.getCaseAndLockByCaseId(caseId);
 
     if (!oCase.isPresent()) {
       throw new RuntimeException(
           "Failed to get row to increment responses, row is probably locked and this should resolve itself: "
-              + caze.getCaseId());
+              + caseId);
     }
 
-    Case lockedCase = oCase.get();
-    lockedCase.setCeActualResponses(lockedCase.getCeActualResponses() + 1);
+    return oCase.get();
+  }
 
-    if (lockedCase.getAddressLevel().equals("U")
-        && lockedCase.getCeActualResponses().intValue()
-            >= lockedCase.getCeExpectedCapacity().intValue()) {
-      lockedCase.setReceiptReceived(true);
-    }
+  @AllArgsConstructor
+  @EqualsAndHashCode
+  @ToString
+  private class Key {
+    private String caseType;
+    private String addressLevel;
+    private String formType;
+  }
+
+  @AllArgsConstructor
+  private class Rule {
```

Comment:
This should really be annotated @Data and us use getters. Disable the setters if you want - Lombok has an option to do that.

I agree that getters and setters are kinda dinosaur idiocy, but it's the convention, so we should follow it.

Reply:
didn't work with @Data, just stuck at explicit getter on there, exhausting